obspy/io/seiscomp/inventory.py

```diff
@@ -111,11 +111,64 @@ def _ns(tagname):
     module = None
     module_uri = None
 
+    # Find the inventory root element. (Only finds the first. We expect only
+    # one, so any more than that will be ignored.)
+    inv_element = root.find(_ns("Inventory"))
+
+    # Pre-generate a dictionary of the sensors, dataloggers and responses to
+    # avoid costly linear search when parsing network nodes later.
+    # Register sensors
+    sensors = {}
+    for sensor_element in inv_element.findall(_ns("sensor")):
+        public_id = sensor_element.get("publicID")
+        if public_id:
+            if public_id in sensors:
+                msg = ("Found multiple matching sensor tags with the same "
+                       "publicID '{}'.".format(public_id))
+                raise obspy.ObsPyException(msg)
+            else:
+                sensors[public_id] = sensor_element
+    # Register dataloggers
+    dataloggers = {}
+    for datalogger_element in inv_element.findall(_ns("datalogger")):
+        public_id = datalogger_element.get("publicID")
+        if public_id:
+            if public_id in dataloggers:
+                msg = ("Found multiple matching datalogger tags with the same "
+                       "publicID '{}'.".format(public_id))
+                raise obspy.ObsPyException(msg)
+            else:
+                dataloggers[public_id] = datalogger_element
+    # Register reponses
+    responses = {
+        "responsePAZ": {},
+        "responsePolynomial": {},
+        "responseFIR": {},
+        "responseIIR": {}
+    }
+    for response_type in responses.keys():
+        for response_element in inv_element.findall(_ns(response_type)):
+            public_id = response_element.get("publicID")
+            if public_id:
+                if public_id in responses[response_type]:
+                    msg = ("Found multiple matching {} tags with the same "
+                           "publicID '{}'.".format(response_type, public_id))
+                    raise obspy.ObsPyException(msg)
+                else:
+                    responses[response_type][public_id] = response_element
```

Comment:
Can be shortened a tiny bit (feel free to choose a different name for `all_elements`).
```python3
    for response_type, all_elements in responses.items():
        for response_element in inv_element.findall(_ns(response_type)):
            public_id = response_element.get("publicID")
            if public_id:
                if public_id in all_elements:
                    msg = ("Found multiple matching {} tags with the same "
                           "publicID '{}'.".format(response_type, public_id))
                    raise obspy.ObsPyException(msg)
                else:
                    all_elements[public_id] = response_element
```

Reply:
Changed as suggested.